cloudplatform/connectivity-oauth/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/OAuth2ServiceImpl.java

```diff
@@ -42,20 +46,49 @@
 @Slf4j
 class OAuth2ServiceImpl
 {
-    private final XsuaaTokenFlows tokenFlows;
-
-    OAuth2ServiceImpl( final XsuaaTokenFlows tokenFlows )
+    /**
+     * Cache to reuse OAuth2TokenService and with that reuse the underlying response cache.
+     * <p>
+     * The {@code OAuth2ServiceImpl} is newly instantiated by {@code OAuth2DestinationBuilder} and
+     * {@code OAuth2ServiceBindingDestinationLoader} for each destination they build/load. This means, without the
+     * cache, also new {@code OAuth2TokenService} would be created for each destination, which in turns defeats the
+     * purpose of the response cache used therein.
+     * <p>
+     * The cache key is composed of the following parts:
+     * <ul>
+     * <li>{@code zoneId}, that way flow is tenant isolated, especially for the HttpClient used against the OAuth2
+     * service.</li>
+     * <li>{@code ClientIdentity}, to separate by the credentials used against the OAuth2 service.</li>
+     * </ul>
+     */
+    private static final Cache<CacheKey, OAuth2TokenService> tokenServiceCache =
+        Caffeine.newBuilder().expireAfterAccess(1, TimeUnit.HOURS).build();
+    private final OAuth2ServiceEndpointsProvider endpoints;
+    private final ClientIdentity identity;
+
+    OAuth2ServiceImpl( final String uri, final ClientIdentity identity )
     {
-        this.tokenFlows = tokenFlows;
+        endpoints = Endpoints.fromBaseUri(URI.create(uri));
+        this.identity = identity;
     }
 
     static OAuth2ServiceImpl fromCredentials( final String uri, final ClientIdentity identity )
     {
-        final DefaultOAuth2TokenService tokenService =
-            new DefaultOAuth2TokenService(HttpClientFactory.create(identity));
-        final OAuth2ServiceEndpointsProvider endpoints = Endpoints.fromBaseUri(URI.create(uri));
-        final XsuaaTokenFlows tokenFlow = new XsuaaTokenFlows(tokenService, endpoints, identity);
-        return new OAuth2ServiceImpl(tokenFlow);
+        return new OAuth2ServiceImpl(uri, identity);
+    }
+
+    static void clearCache()
+    {
+        log.warn("Resetting the TokenService cache. This should not be done outside of testing.");
+        tokenServiceCache.invalidateAll();
+    }
+
+    XsuaaTokenFlows createTokenFlow( final String zoneId )
```

Comment:
(Major)

To avoid future confusion, this method does not "create a token flow" but it "creates a token flow factory" 😁 

Maybe...
```suggestion
    XsuaaTokenFlows getTokenFlows( final String zoneId )
```